static/js/services/roompin.js

```diff
@@ -23,40 +23,95 @@
 define([
 ], function() {
 
-	return ["$window", "$q", "alertify", "translation", "safeMessage", function($window, $q, alertify, translation, safeMessage) {
+	return ["$window", "$q", "globalContext", "alertify", "translation", "safeMessage", "randomGen", "localStorage", function($window, $q, context, alertify, translation, safeMessage, randomGen, localStorage) {
 
 		var pinCache = {};
+		var getLocalStoragePINIDForRoom = function(roomName) {
+			return "room-pin-" + roomName;
+		};
+		var lockedRoomsJoinable = !!context.Cfg.LockedRoomJoinableWithPIN;
 		var roompin = {
 			get: function(roomName) {
 				var cachedPIN = pinCache[roomName];
 				return cachedPIN ? cachedPIN : null;
 			},
 			clear: function(roomName) {
 				delete pinCache[roomName];
+				localStorage.removeItem(getLocalStoragePINIDForRoom(roomName));
 				console.log("Cleared PIN for", roomName);
 			},
 			update: function(roomName, pin) {
 				if (pin) {
 					pinCache[roomName] = pin;
-					alertify.dialog.alert(translation._("PIN for room %s is now '%s'.", safeMessage(roomName), safeMessage(pin)));
+					localStorage.setItem(getLocalStoragePINIDForRoom(roomName), pin);
+					if (lockedRoomsJoinable) {
+						alertify.dialog.alert(translation._("PIN for room %s is now '%s'.", safeMessage(roomName), safeMessage(pin)));
+					}
 				} else {
 					roompin.clear(roomName);
-					alertify.dialog.alert(translation._("PIN lock has been removed from room %s.", safeMessage(roomName)));
 				}
 			},
 			requestInteractively: function(roomName) {
 				var deferred = $q.defer();
-				alertify.dialog.prompt(translation._("Enter the PIN for room %s", safeMessage(roomName)), function(pin) {
+				var tryJoinWithStoredPIN = function() {
+					var pin = localStorage.getItem(getLocalStoragePINIDForRoom(roomName));
 					if (pin) {
-						pinCache[roomName] = pin;
+						roompin.update(roomName, pin);
 						deferred.resolve();
-					} else {
+						return true;
+					}
+					return false;
+				};
+				if (lockedRoomsJoinable) {
+					if (!tryJoinWithStoredPIN()) {
+						alertify.dialog.prompt(translation._("Enter the PIN for room %s", safeMessage(roomName)), function(pin) {
+							if (pin) {
+								roompin.update(roomName, pin);
+								deferred.resolve();
+							} else {
+								deferred.reject();
+							}
+						}, function() {
+							deferred.reject();
+						});
+					}
+				} else {
+					if (!tryJoinWithStoredPIN()) {
+						alertify.dialog.error(
+							translation._("Can't join locked room '%s'.", safeMessage(roomName)),
+							translation._("Room '%s' is locked. This server is configured to not let anyone join locked rooms.", safeMessage(roomName))
+						);
 						deferred.reject();
 					}
+				}
+				return deferred.promise;
+			},
+			// This is bit of a hack to have all room-PIN related stuff at one place
```

Comment:
Well, it's a hack because we need to pass `rooms` to prevent circular dependency.